### PR TITLE
[Fixes #1184] Formatting issues in requests should not cause a 500 respo...

### LINKED ISSUE #1184: Formatting issues in requests should not cause a 500 response
Currently we return a 500 response if there are any formatting issues with the incoming request (ex bad json, incorrect json, etc). We should instead capture these errors as validation errors and let the app choose to either return a 400 or ignore the errors.

EDIT: Clarification the result should be a ModelState error and ModelState.IsValid == false

### PR DESCRIPTION
**Pull request by kichalla**
...nse

@yishaigalatzer @rynowak @harshgMSFT

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs [hunk 1]
```diff
@@ -58,8 +59,28 @@ public class BodyModelBinder : BindingSourceModelBinder
...
 
-            var model = await formatter.ReadAsync(formatterContext);
+            object model = null;
+            try
+            {
+                model = await formatter.ReadAsync(formatterContext);
+            }
+            catch (Exception ex)
+            {
+                model = GetDefaultValueForType(bindingContext.ModelType);
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: FYI: following a similar approach as in Web API
Example:
`ReadFromStream` method in
http://aspnetwebstack.codeplex.com/SourceControl/latest#src/System.Net.Http.Formatting/Formatting/XmlMediaTypeFormatter.cs

File: test/WebSites/XmlFormattersWebSite/Controllers/HomeController.cs [hunk 2]
```diff
@@ -13,7 +14,52 @@ public class HomeController : Controller
...
+                    StatusCode = 400
+                };
+            }
+
             return Content(dummyObject.SampleInt.ToString());
         }
+
+        // Cannot use 'SerializableError' here as it sanitizes exceptions in model state with generic error message.
+        // Since the tests need to verify the messages, we are doing the following.
+        private List<string> GetModelStateErrorMessages(ModelStateDictionary modelStateDictionary)
```
Comment (kichalla) [inline test/WebSites/XmlFormattersWebSite/Controllers/HomeController.cs]: FYI: This code is present in my other PR too. I will remove this and probably move it to a common place and use it.

File: src/Microsoft.AspNet.Mvc.Core/Formatters/JsonInputFormatter.cs [hunk 3]
```diff
@@ -79,19 +73,17 @@ public override Task<object> ReadRequestBodyAsync([NotNull] InputFormatterContex
                 var jsonSerializer = CreateJsonSerializer();
 
                 EventHandler<Newtonsoft.Json.Serialization.ErrorEventArgs> errorHandler = null;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/JsonInputFormatter.cs]: Add using if necessary

File: src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs [hunk 4]
```diff
@@ -58,8 +59,28 @@ public class BodyModelBinder : BindingSourceModelBinder
...
+                bindingContext.ModelState.AddModelError(bindingContext.ModelName, ex);
+            }
+
             return new ModelBindingResult(model, bindingContext.ModelName, isModelSet: true);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs]: Should we return something different when this throws?

**Comment by rynowak** [comment]
:shipit: after a quick discussion about the feedback

File: src/Microsoft.AspNet.Mvc.Core/Formatters/JsonInputFormatter.cs (hunk 3 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/JsonInputFormatter.cs]: Here ErrorEventArgs is conflicting with the type from `System.IO` namespace, so fully qualifying it here.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/JsonInputFormatter.cs]: :ok:

**Comment by kichalla** [comment]
Updated. @rynowak

**Pull request merged**